clippy_lints/src/type_param_mismatch.rs

```diff
@@ -0,0 +1,113 @@
+        if_chain! {
+            if !item.span.from_expansion();
+            if let ItemKind::Impl(imp) = &item.kind;
+            if let TyKind::Path(QPath::Resolved(_, path)) = &imp.self_ty.kind;
+            if let Some(segment) = path.segments.iter().next();
+            if let Some(generic_args) = segment.args;
+            if !generic_args.args.is_empty();
+            then {
+                // get the name and span of the generic parameters in the Impl
+                let impl_params = generic_args.args.iter()
+                .filter_map(|p|
+                    match p {
+                        GenericArg::Type(Ty {kind: TyKind::Path(QPath::Resolved(_, path)), ..}) =>
+                            Some((path.segments[0].ident.to_string(), path.span)),
+                        _ => None,
+                    }
+                );
+
+                // find the type that the Impl is for
+                // only lint on struct/enum/union for now
+                let defid = match path.res {
+                    Res::Def(DefKind::Struct | DefKind::Enum | DefKind::Union, defid) => defid,
+                    _ => return,
+                };
+
+                let type_name = segment.ident;
+
+                // get the names of the generic parameters in the type
+                let type_params = &cx.tcx.generics_of(defid).params;
+                let type_param_names: Vec<_> = type_params.iter()
+                .filter_map(|p|
+                    match p.kind {
+                        GenericParamDefKind::Type {..} => Some(p.name.to_string()),
+                        _ => None,
+                    }
+                ).collect();
+
+                for (i, (impl_param_name, impl_param_span)) in impl_params.enumerate() {
+                    if mismatch_param_name(i, &impl_param_name, &type_param_names) {
+                        let msg = format!("this type parameter is named {}, but it does not refer to {}::{}",
+                                  impl_param_name, type_name, impl_param_name);
+                            span_lint_and_sugg(
+                                cx,
+                                TYPE_PARAM_MISMATCH,
+                                impl_param_span,
+                                &msg,
+                                "try",
+                                type_param_names[i].to_string(),
+                                Applicability::MaybeIncorrect
+                            );
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Checks if impl_param_name is the same as one of type_param_names,
+// and is in a different position
+fn mismatch_param_name(i: usize, impl_param_name: &str, type_param_names: &[String]) -> bool {
+    for (j, type_param_name) in type_param_names.iter().enumerate() {
+        if impl_param_name == type_param_name && i != j {
+            return true;
+        }
+    }
+    false
+}
```

Comment:
Instead of iterating over `type_param_names` and checking the index, we can store `type_param_names` in a hashmap of `name => index` and just fetch the value from there.

Reply:
I considered a hashmap, but it seems a bit unnecessary/more loc. The number of things we are iterating through is probably low. 

Willing to change if you prefer, though.

Reply:
I prefer using a hashmap, it makes the code easier to understand, and I don't see any downside from it.